Add invalid checks for static ranges

jesxilin opened the pull request:
#### 8afe94ae85ece85e0453b8cdf574b3c23b01c800
<pre>
Add invalid checks for static ranges
<a href="https://bugs.webkit.org/show_bug.cgi?id=258826">https://bugs.webkit.org/show_bug.cgi?id=258826</a>
rdar://111706100

Reviewed by Ryosuke Niwa.

Added invalid checks for static range.
Thus, ignoring the invalid range when painting highlights.
Changed TestExpectations due to more tests passing.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/StaticRange.cpp:
(WebCore::StaticRange::computeRangeValidity const):
(WebCore::StaticRange::isValidForPaintingHighlights const):
* Source/WebCore/dom/StaticRange.h:
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForHighlights):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::coalesceAdjacentWithSameRanges):

Canonical link: <a href="https://commits.webkit.org/266019@main">https://commits.webkit.org/266019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/510a055f58826f848172d3083b5402c656a2a931

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12576 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13220 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14316 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/12059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12637 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15407 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12923 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14747 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12740 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13517 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10639 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14760 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10795 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11382 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11879 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11547 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/14751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12032 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9947 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11273 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15606 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1430 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11888 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,14 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
 
+bool StaticRange::isValid() const
+{
+    unsigned containerLength = startContainer().length() + endContainer().length();
```
This may overflow and we should probably return false if it does, no?

See `CheckedUint32` type to deal with overflows.

cdumez reviewed (commented):
Is this testable?

aprotyas commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,14 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+
+    if (endOffset() >= startOffset())
+        return (endOffset() - startOffset()) <= containerLength;
+    return (startOffset() - endOffset()) <= containerLength;
```
Nit: `std::abs`?

jesxilin replied:
I had that at first but apparently it could overflow even if we casted it to int?

aprotyas replied:
Oh maybe I didn't think it all the way through. Feel free to keep it as-is!

aprotyas commented on LayoutTests/TestExpectations:
```diff
@@ -4918,8 +4918,7 @@ imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-
 imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-invalidation-006.html [ ImageOnlyFailure ]
-imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-staticrange-001.html [ ImageOnlyFailure ]
- imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-017.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-017.html [ ImageOnlyFailure ]
```
Are there any other expectations we can adjust?

jesxilin replied:
I was thinking of doing that in a different pr but I can check it now and include it in this one.

aprotyas commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,12 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+{
+    if ((endOffset() < startOffset() && startContainer().isSameNode(&endContainer())) || collapsed() || startContainer().isDocumentNode() || endContainer().isDocumentNode() || startContainer().isDocumentTypeNode() || endContainer().isDocumentTypeNode()
+        || !connectedInSameTreeScope(&startContainer().rootNode(), &endContainer().rootNode())
+        || is_gt(treeOrder<ComposedTree>(startContainer(), endContainer())))
```
Hmm... we should probably reorganize this section

aprotyas replied:
How about something like this?

```cpp
bool StaticRange::isValid() const
{
    if (collapsed())
        return false;
    if (endOffset() < startOffset() && startContainer().isSameNode(&endContainer()))
        return false;
    if (startContainer().isDocumentNode() || startContainer().isDocumentTypeNode())
        return false;
    if (endContainer().isDocumentNode() || endContainer().isDocumentTypeNode())
        return false;
    if (!connectedInSameTreeScope(&startContainer().rootNode(), &endContainer().rootNode()))
        return false;
    if (is_gt(treeOrder<ComposedTree>(startContainer(), endContainer())))
        return false;
    return endOffset() <= endContainer().length() && startOffset() <= startContainer().length();
}
```

cdumez replied:
For what it's worth, I find @aprotyas 's proposal more readable too.

jesxilin replied:
will do 👍

rr-codes replied:
you could also probably cache `endContainer` and `startContainer` into local variables

rr-codes commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -122,6 +122,8 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
                 for (auto& rangeData : highlightRegister->map().get(highlightName)->rangesData()) {
                     if (!highlightData.setRenderRange(rangeData))
                         continue;
+                    if (auto* staticRange = dynamicDowncast<StaticRange>(rangeData->range()); staticRange && !staticRange->isValid())
```
should this be
```suggestion
                    if (auto* staticRange = dynamicDowncast<StaticRange>(rangeData->range()); !staticRange || !staticRange->isValid())
```

(ie, should it `continue` if the range is not a StaticRange?)

jesxilin replied:
It should not `continue` if range is not static range bc we support live ranges too. I think you're suggesting to `continue` if it's not static range?

rr-codes replied:
ah I see, just wanted to confirm!

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,23 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+
+    if (collapsed())
+        return false;
+    if (endOffset() < startOffset() && startContainer.isSameNode(&endContainer))
```
Just do: &startContainer == &endContainer
isSameNode only exists for web backwards compatibility reasons.

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,23 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
     addWebCoreOpaqueRoot(visitor, end.container.get());
 }
 
+bool StaticRange::isValid() const
```
Please add a link to https://dom.spec.whatwg.org/#staticrange-valid as a comment.

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,23 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+        return false;
+    if (endOffset() < startOffset() && startContainer.isSameNode(&endContainer))
+        return false;
+    if (startContainer.isDocumentNode() || startContainer.isDocumentTypeNode())
```
Where did this check come from. It's not listed in https://dom.spec.whatwg.org/#staticrange-valid.

jesxilin replied:
There is a WPT `imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-staticrange-001.html`. One case has `document` as the container and it is deemed invalid since the first child of `document` would be `<!DOCTYPE html>` which doesn't make sense to style. So nothing will/should be painted.

rniwa replied:
Have you considered the possibility that WPT is wrong?

https://dom.spec.whatwg.org/#staticrange-valid doesn't check this condition and there is nothing regarding document node or document type node in https://drafts.csswg.org/css-highlight-api-1/#range-invalidation

Either we're missing some text in the spec, which should be filed as a spec issue, or that particular test is wrong.

jesxilin replied:
I asked people from web standards and they also said it could potentially be wrong but the logic would be to grab `<!DOCTYPE html>` in which you can't highlight that. Thus, don't paint. They said we could discuss it more to determine if the test is wrong but so far Chrome and Firefox both pass this test as well.

rniwa replied:
Not painting makes sense. What doesn't is making this isValid concept depart from the definition in the spec.

I think we should get rid of these checks from isValid and placed them in where isValid is called instead.

Alternatively, we should add a comment saying this doesn't match the spec's definition of isValid and add a URL to a spec issue.

rniwa replied:
Put it differently, conceptually, there is a concept of static range being valid vs. static range being paintable as highlights. It totally makes sense that a static range anchored at doctype can't be painted but that's a different concept from static range being valid, which has a very specific definition.

Also, why is anchoring node on document node wrong / invalid? If we specified `display: block` on `<html>` element, it will be rendered in which case highlight would be paintable at least in concept.

jesxilin replied:
For the case you mentioned, I guess it _would_ be paintable but I don't think it _should_? I wouldn't know what the styling would apply to? I'll ask to make sure of this.

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,23 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+        return false;
+    if (!connectedInSameTreeScope(&startContainer.rootNode(), &endContainer.rootNode()))
+        return false;
+    if (is_gt(treeOrder<ComposedTree>(startContainer, endContainer)))
```
This is a O(n) operation. We may want to consider calling this function computeValidity instead to signify the expensive nature of this function call.

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,31 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+
+    if (collapsed())
+        return false;
+    if (is_gt(treeOrder<ComposedTree>(startContainer, endContainer)))
```
This check should probably be a part of StaticRange::isValid since the definition of being valid includes that condition:
https://dom.spec.whatwg.org/#staticrange-valid

jesxilin replied:
The `if (is_gt(treeOrder<ComposedTree>(startContainer, endContainer)))`?

rniwa replied:
Yes. "Its [start](https://dom.spec.whatwg.org/#concept-range-start) is [before](https://dom.spec.whatwg.org/#concept-range-bp-before) or [equal](https://dom.spec.whatwg.org/#concept-range-bp-equal) to its [end](https://dom.spec.whatwg.org/#concept-range-end)." corresponds to that check.

rniwa commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -122,6 +122,20 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                        && (!staticRange->isValid() || !staticRange->computeValidityForPaintingHighlights()))
+                        continue;
+                    // FIXME: Potentially move this check elsewhere, to where we collect this range information.
+                    bool noRenderer = true;
```
It's probably more readable to wrap this code in an anonymous lambda like this:
```
auto hasRenderer = [&]() {
    IntersectingNodeRange nodes(makeSimpleRange(rangeData->range()));
    for (auto& iterator : nodes) {
        if (iterator.renderer())
            return true;
    }
    return false;
}();
```

jesxilin replied:
Will make this change soon! Thanks, it def looks much better this way.

rr-codes commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -122,6 +122,20 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                        && (!staticRange->computeValidity() || staticRange->collapsed()))
+                        continue;
+                    // FIXME: Potentially move this check elsewhere, to where we collect this range information.
+                    auto hasRenderer = [&]() {
```
```suggestion
                    auto hasRenderer = [&] {
```

(very very optional change)

aprotyas commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -147,10 +147,10 @@ static Vector<StyledMarkedText> coalesceAdjacentWithSameRanges(Vector<StyledMark
                 previousStyledMarkedText.priority = it->priority;
                 // If highlight, take textDecorationStyles of latest StyledMarkedText.
-                // FIXME: Check for taking textDecorationStyles needs to be changed to accommodate other MarkedText type
+                // FIXME: Check for taking textDecorationStyles needs to be changed to accommodate other MarkedText type.
```
Is this a "should be on bugzilla" type FIXME?

jesxilin replied:
what do you mean?

jesxilin replied:
got it!

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,21 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+    if (is_gt(treeOrder<ComposedTree>(startContainer, endContainer)))
+        return false;
+    // Its start and end are in the same node tree.
+    if (!connectedInSameTreeScope(&startContainer.rootNode(), &endContainer.rootNode()))
```
We should do this check

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,21 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+        return false;
+    // Its start offset is between 0 and its start node’s length, inclusive.
+    // Its end offset is between 0 and its end node’s length, inclusive.
+    return endOffset() <= endContainer.length() && startOffset() <= startContainer.length();
```
And this check before checking is_gt(treeOrder<ComposedTree>(...), which is O(n).

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,24 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+    Node& endContainer = this->endContainer();
+
+    // Its start is before or equal to its end, if same container.
+    if (endOffset() < startOffset() && &startContainer == &endContainer)
```
This check should probably appear immediately above is_gt since the relevant clause in the spec here is "Its [start](https://dom.spec.whatwg.org/#concept-range-start) is [before](https://dom.spec.whatwg.org/#concept-range-bp-before) or [equal](https://dom.spec.whatwg.org/#concept-range-bp-equal) to its [end](https://dom.spec.whatwg.org/#concept-range-end)."

By the way, these "what" comments are unnecessary. They just repeat what the code already conveys.

jesxilin replied:
Will make this change and take out the comments.

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,24 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+    if (startOffset() > startContainer.length() || startOffset() < 0)
+        return false;
+    // Its end offset is between 0 and its end node’s length, inclusive.
+    if (endOffset() > endContainer.length() || endOffset() < 0)
```
Ditto.

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,24 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+    if (!connectedInSameTreeScope(&startContainer.rootNode(), &endContainer.rootNode()))
+        return false;
+    // Its start offset is between 0 and its start node’s length, inclusive.
+    if (startOffset() > startContainer.length() || startOffset() < 0)
```
Since `startOffset()` is unsigned, `startOffset() < 0` is a redundant check.

rniwa commented on Source/WebCore/dom/StaticRange.cpp:
```diff
@@ -83,4 +83,19 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
+        return false;
+    if (endOffset() > endContainer.length())
+        return false;
+    if (endOffset() < startOffset() && &startContainer == &endContainer)
```
I guess this could be just:
```
if (&startContainer == &endContainer)
    return endOffset() < startOffset();
```

jesxilin replied:
Logically yes. Will try, If it works I'll push it. Believe it should be `if (&startContainer == &endContainer)
    return endOffset() > startOffset();`?

jesxilin committed:
Add invalid checks for static ranges
https://bugs.webkit.org/show_bug.cgi?id=258826
rdar://111706100

Reviewed by Ryosuke Niwa.

Added invalid checks for static range.
Thus, ignoring the invalid range when painting highlights.
Changed TestExpectations due to more tests passing.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/StaticRange.cpp:
(WebCore::StaticRange::computeRangeValidity const):
(WebCore::StaticRange::isValidForPaintingHighlights const):
* Source/WebCore/dom/StaticRange.h:
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForHighlights):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::coalesceAdjacentWithSameRanges):

Canonical link: https://commits.webkit.org/266019@main
